### PR TITLE
introduce a "missingness" option/flag on the VCF class

### PR DESCRIPTION
**Pull request by indraniel**
This pull request introduces an optional "missingness" boolean flag feature when iterating over a VCF.  

When the missingness flag is enabled, `cyvcf2` will treat any genotype containing a missing allele (containing a '.') as an UNKNOWN genotype; otherwise genotypes like `0/.`, `./0`, `1/.`, or `./1` will be classified as heterozygous ("HET").

To activate the feature, one can simply do:

```python
from cyvcf2 import VCF
vcf = VCF("/path/to/vcf/file", missingness=True)

for variant in vcf:
    # do something
```

The "missingness" feature is not enabled by default (i.e. set to `False` by default) to preserve backward compatibility.

This option was created to address the issue described in [`cyvcf2`'s github issue 54][0].

[0]: https://github.com/brentp/cyvcf2/issues/54

### CODE REVIEW

**Comment by brentp** [comment]
this looks nicely done. can you run through a large VCF with > 10 samples, accessing gt_types on eahc variant and compare the  speed before and after this change?

I'm worried about performance changes and adding extra surface area to the API.

**Comment by indraniel** [comment]
Sure thing!  How many variant sites (vcf rows) should there approximately be for this "large" test VCF?

**Comment by brentp** [comment]
not sure. 10's of thousands at least, to get an idea of the change.

**Comment by indraniel** [comment]
I finally got around to making the requested benchmark.  It appears the new missingness feature makes the code slightly slower for VCF files with a small number of variants, but has comparatively similar timings for VCF files containing a large number of variants (> 1,000,000 variants).

Here are some benchmarking plots:
![Benchmarking Execution Times (Linear Scale)][0]
![Benchmarking Execution Times (Log Scale)][1]

The benchmarking methodology and data details are available at [indraniel/cyvcf2-missingness-feature-benchmarks][2]. 

Feel free to let me know if you find anything unclear, or if any further work is needed.

[0]: https://github.com/indraniel/cyvcf2-missingness-feature-benchmarks/blob/master/results/benchmark.png
[1]: https://github.com/indraniel/cyvcf2-missingness-feature-benchmarks/blob/master/results/benchmark-log-scale.png
[2]: https://github.com/indraniel/cyvcf2-missingness-feature-benchmarks

File: cyvcf2/cyvcf2.pyx [hunk 1]
```diff
@@ -950,7 +954,7 @@ cdef class Variant(object):
             cdef int unknown = 3 if self.vcf.gts012 else 2
             for i in range(0, n * self.vcf.n_samples, n):
                 if n == 2:
-                    if gt_types[j] == unknown:
+                    if (gt_types[j] == unknown) and (self.vcf.missingness == False):
```
Comment (brentp) [inline cyvcf2/cyvcf2.pyx]: can you change to `and (not self.vcf.missingness)` ?

**Comment by brentp** [comment]
thanks very much for doing that. It looks like it's a negligible difference and well worth it.
Other than what I mentioned in the review, it looks good.

File: cyvcf2/cyvcf2.pyx (hunk 1 above)
Comment (indraniel) [inline cyvcf2/cyvcf2.pyx]: sure thing!  Just pushed [the change][0].

[0]: https://github.com/brentp/cyvcf2/pull/55/commits/79173edeec45c3811b573be3b26f92ffa6a6ee97

Comment (indraniel) [inline cyvcf2/cyvcf2.pyx]: Yeah.  I don't like the term `missingness` either, but I've still yet to think of a better term for this feature.

**Comment by indraniel** [comment]
Yeah. I don't like the name `missingness` either, but I've still yet to think of a better term for this feature.

**Comment by brentp** [comment]
what about `strict_missing_gt`?

**Comment by indraniel** [comment]
`strict_missing_gt` is better...can we go even more extreme and just say `strict_gt` ?

**Comment by brentp** [comment]
yes please.

**Comment by indraniel** [comment]
I just updated the logic using the new term of `strict_gt` instead of `missingness`.  Let me know if any further modifications are needed.

**Comment by indraniel** [comment]
I updated the documentation page to mention the new "genotyping strictness" feature as well.

**Comment by brentp** [comment]
Looks good to me. Is this ready to go?

**Comment by indraniel** [comment]
I'm ready to release it, once the checks pass.

**Comment by indraniel** [comment]
Looks like things are ready to go!

**Comment by brentp** [comment]
cheers

**Pull request merged**